Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -0,0 +1,285 @@
+  // extend beyond `Write` (the buffer must be valid until the completion queue
+  // returns the tag associated with the write, see
+  // https://github.com/grpc/grpc/issues/13019#issuecomment-336932929, #5).
+  grpc::ByteBuffer message_;
+};
+
+class ServerInitiatedFinish : public StreamOperation {
+ public:
+  using StreamOperation::StreamOperation;
+
+  void DoExecute(grpc::GenericClientAsyncReaderWriter* call) override {
+    call->Finish(&grpc_status_, this);
+  }
+
+ private:
+  void DoComplete() override {
+    // Note: calling Finish on a GRPC call should never fail, according to the
+    // docs
+    delegate_.OnFinishedWithServerError(grpc_status_);
+  }
+
+  grpc::Status grpc_status_;
+};
+
+class ClientInitiatedFinish : public StreamOperation {
+ public:
+  using StreamOperation::StreamOperation;
+
+ private:
+  void DoExecute(grpc::GenericClientAsyncReaderWriter* call) override {
+    call->Finish(&unused_status_, this);
+  }
+
+  void DoComplete() override {
+    // Nothing to do
+  }
+
+  grpc::Status unused_status_;
+};
+
+}  // namespace
+
+// Call
+
+GrpcStream::GrpcStream(
+    std::unique_ptr<grpc::ClientContext> context,
+    std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+    GrpcOperationsObserver* observer,
+    GrpcCompletionQueue* grpc_queue)
+    : context_{std::move(context)},
+      call_{std::move(call)},
+      observer_{observer},
+      grpc_queue_{grpc_queue},
+      generation_{observer->generation()} {
+}
+
+void GrpcStream::Start() {
+  HARD_ASSERT(!is_started_, "Call is already started");
+  is_started_ = true;
+  Execute<StreamStart>();
+}
+
+void GrpcStream::Read() {
+  HARD_ASSERT(!has_pending_read_,
+              "Cannot schedule another read operation before the previous read "
+              "finishes");
+  if (write_and_finish_) {
+    return;
+  }
+
+  has_pending_read_ = true;
+  Execute<StreamRead>();
+}
+
+void GrpcStream::Write(grpc::ByteBuffer&& message) {
+  if (write_and_finish_) {
+    return;
+  }
+
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::Finish() {
+  buffered_writer_.reset();
+  context_->TryCancel();
+  Execute<ClientInitiatedFinish>();
+}
+
+void GrpcStream::WriteAndFinish(grpc::ByteBuffer&& message) {
+  if (!buffered_writer_) {
+    // Ignore the write part if the call didn't have a chance to open yet.
+    Finish();
+    return;
+  }
+
+  write_and_finish_ = true;
+  // Write the last message as soon as possible by discarding anything else that
+  // might be buffered.
+  buffered_writer_->DiscardUnstartedWrites();
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::BufferedWrite(grpc::ByteBuffer&& message) {
+  std::unique_ptr<StreamWrite> write_operation{
+      MakeOperation<StreamWrite>(std::move(message))};
+  buffered_writer_->Enqueue(std::move(write_operation));
+}
+
+bool GrpcStream::SameGeneration() const {
+  return generation_ == observer_->generation();
```

Comment:
FYI- This ended up getting named closeCount in JS / Android and I ended up creating a closeGuardedDispatcher / closeGuardedRunner to do the check.  If it's convenient to move this in that direction for porting symmetry, that would be nice.

Also, I haven't traced through this to figure out where we dispatch onto our queue (if gRPC is doing it for us or we're doing it explicitly somewhere) but in JS, CloseGuardedDispatcher takes care of that as well. And in particular it's critical that the SameGeneration() check happens _after_ we're on our queue. And so if nothing else, I think I'd expect to see a `VerifyIsCurrentQueue()` call in here.

Also, unless I'm missing it, you're not incrementing generation currently.  That needs to happen on close somehow...